AdyenDropIn/Modules/Component/ComponentContainerViewModel.swift

```diff
@@ -61,17 +94,79 @@ internal class ComponentContainerViewModel: ComponentContainerViewModelProtocol
 extension ComponentContainerViewModel: PaymentComponentDelegate {
 
     func didSubmit(
-        _ data: Adyen.PaymentComponentData,
-        from component: any Adyen.PaymentComponent
+        _ data: PaymentComponentData,
+        from component: any PaymentComponent
     ) {
-        // TODO: -
-        print("SUBMIT")
+        let checkoutAttemptId = component.context.analyticsProvider?.checkoutAttemptId
+        let updatedData = data.replacing(
+            checkoutAttemptId: checkoutAttemptId
+        )
+
+        guard updatedData.browserInfo == nil else {
+            delegate?.didSubmit(updatedData, from: component)
+            return
+        }
+        updatedData.dataByAddingBrowserInfo { [weak self] in
+            guard let self else { return }
+            delegate?.didSubmit($0, from: component)
+        }
     }
     
     func didFail(
         with error: any Error,
         from component: any Adyen.PaymentComponent
     ) {
-        // TODO: -
+        if case ComponentError.cancelled = error {
+            handleCancellation()
+        } else {
+            delegate?.didFail(with: error)
+        }
+    }
+
+    // MARK: - Private
+
+    private func handleCancellation() {
+        component.cancelIfNeeded()
```

Comment:
can we simplify and call it `cancel`? on the API level it adds unnecessary context

Reply:
Move this logic to a `cancel` method that just calling `handleCancellation `.